ldj-spring-renderer-example/src/main/java/com/infosupport/ldoc/springexample/SpringEventRenderer.java

```diff
@@ -0,0 +1,173 @@
+
+    if (node.path("$type").asText("").contains("InvocationDescription")
+        && node.path("ContainingType").asText("").startsWith(CONTEXT_PACKAGE)
+        && node.path("Name").asText("").equals("publishEvent")) {
+      // This is an explicit call to some Spring Context 'publishEvent' method.
+      String eventType = node.get("Arguments").get(0).get("Type").textValue();
+      interactions.add(new Interaction(Kind.PUBLISHES, sendingClass, eventType, eventName));
+    }
+
+    for (JsonNode child : node) {
+      interactions.addAll(findPublishCalls(sendingClass, eventName, child));
+    }
+
+    return interactions;
+  }
+
+  static List<Interaction> findInteractions(JsonNode root) {
+    List<Interaction> interactions = new ArrayList<>(0);
+
+    for (JsonNode type : root) {
+      String fullName = type.path("FullName").asText("");
+
+      for (JsonNode method : type.path("Methods")) {
+        String returnType = method.get("ReturnType").asText();
+        String eventType = null;
+
+        for (JsonNode annotation : method.path("Attributes")) {
+          eventType = method.path("Parameters").path(0).path("Type").asText(null);
+
+          if (annotation.get("Type").asText().equals(EV_LISTENER)) {
+            // This method is annotated @EventListener, so it is a listener.
+            interactions.add(new Interaction(Kind.LISTENS, fullName, eventType, null));
+
+            if (!returnType.equals("void")) {
+              // It returns something, which is published as a new event, so it's also a publisher.
+              interactions.add(new Interaction(Kind.PUBLISHES, fullName, returnType, eventType));
+            }
+          }
+        }
+
+        // Look through the method for calls to publishEvent, also if the method is not a listener.
+        interactions.addAll(findPublishCalls(fullName, eventType, method.path("Statements")));
+      }
+    }
+
+    interactions.sort(Comparator.comparing(Interaction::className));
+
+    return interactions;
+  }
+
+  /**
+   * Converts a fully-qualified class name into a plain class name.
+   */
+  static String stripName(String fqdn) {
+    return fqdn.replaceAll("^.*\\.", "");
+  }
+
+  /**
+   * Converts a fully-qualified class to a label that seems more human, with space-separated words.
+   */
+  static String humanizeName(String fqdn) {
+    return String.join(" ", stripName(fqdn).split("(?<=\\p{Ll})(?=\\p{Lu})"));
+  }
+
+  static void renderParticipant(PrintWriter out, String name) {
+    out.printf("participant \"%s\" as %s\n", humanizeName(name), name);
+  }
+
+  static void renderInteraction(PrintWriter out, String sender, String receiver, String event) {
+    out.printf("%s -[#ForestGreen]> %s : %s\n", sender, receiver, stripName(event));
+  }
+
+  static void renderInteractions(PrintWriter out, String sender, List<Interaction> interactions,
+      List<String> classes, String scope) {
+    // This is O(spicy) but seems fast enough in our example.
```

Comment:
I would say performance is not an issue. The only thing is the readability for people that try to follow the code (considering the exemplifying nature of this renderer)